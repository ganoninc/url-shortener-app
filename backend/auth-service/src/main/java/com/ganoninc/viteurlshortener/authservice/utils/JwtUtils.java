package com.ganoninc.viteurlshortener.authservice.utils;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;
import java.nio.charset.StandardCharsets;
import java.util.Date;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

@Component
public class JwtUtils {

  @Value("${jwt.secret}")
  private String secret;

  @Value("${jwt.expirationMs}")
  private Long expiration;

  public String generateToken(String email) {
    return Jwts.builder()
        .setSubject(email)
        .setIssuer("auth-service")
        .setIssuedAt(new Date())
        .setExpiration(new Date(System.currentTimeMillis() + expiration))
        .signWith(
            Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8)), SignatureAlgorithm.HS256)
        .compact();
  }
}
