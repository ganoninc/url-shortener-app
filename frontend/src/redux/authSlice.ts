import { createSlice, type PayloadAction } from "@reduxjs/toolkit";
import type { RootState } from "./store";

type User = {
  email: string;
};

export type AuthState =
  | {
      jwt: null;
      user: null;
    }
  | {
      jwt: string;
      user: User;
    };

const initialState: AuthState = {
  jwt: null,
  user: null,
};

const authSlice = createSlice({
  name: "auth",
  initialState,
  reducers: {
    setCredentials: (state: AuthState, action: PayloadAction<AuthState>) => {
      state.jwt = action.payload.jwt;
      state.user = action.payload.user;
    },
    logout: (state: AuthState) => {
      state.jwt = null;
      state.user = null;
    },
  },
});

export const { setCredentials, logout } = authSlice.actions;
export const selectUser = (state: RootState) => state.auth.user;
export const selectJwt = (state: RootState) => state.auth.jwt;
export default authSlice.reducer;
