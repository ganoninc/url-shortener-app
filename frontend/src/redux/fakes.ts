import type { AuthenticatedAuthState } from "./authSlice";

export const fakeAuthenticatedAuthState: AuthenticatedAuthState = {
  status: "authenticated",
  jwt: "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0QHRlc3QuY29tIiwiaWF0IjoxNzAwMDAwMDAwfQ.dGVzdC1zaWduYXR1cmUtbm90LXZhbGlk",
  user: { email: "test@test.com" },
};
