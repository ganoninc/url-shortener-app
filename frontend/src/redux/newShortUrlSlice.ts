import { createSlice, type PayloadAction } from "@reduxjs/toolkit";
import type { RootState } from "@reduxjs/toolkit/query";

export type NewShortUrlState = {
  originalUrl: string;
};

const initialState: NewShortUrlState = {
  originalUrl: "",
};

const newShortUrlSlice = createSlice({
  name: "newShortUrl",
  initialState,
  reducers: {
    updateOriginalUrl: (
      state: NewShortUrlState,
      action: PayloadAction<NewShortUrlState>
    ) => {
      state.originalUrl = action.payload.originalUrl;
    },
  },
});

export const { updateOriginalUrl } = newShortUrlSlice.actions;
export const selectOriginalUrl = (state: RootState) =>
  state.newShortUrl.originalUrl;
export default newShortUrlSlice.reducer;
