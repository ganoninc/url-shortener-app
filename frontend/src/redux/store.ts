import { combineReducers, configureStore } from "@reduxjs/toolkit";
import authReducer from "./authSlice";

const rootReducer = combineReducers({
  auth: authReducer,
});

export function setupStore(preloadedState?: Partial<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof configureStore>;
export type AppDispatch = AppStore["dispatch"];
