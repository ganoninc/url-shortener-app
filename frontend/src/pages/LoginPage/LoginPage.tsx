import { setCredentials } from "../../redux/authSlice";
import Title from "../../components/Title/Title";
import { useAppDispatch } from "../../hooks/hooks";
import Button from "../../components/Button/Button";
import Page from "../../components/Page/Page";

export default function LoginPage() {
  const dispatch = useAppDispatch();
  const API_GATEWAY_URL = import.meta.env.VITE_API_GATEWAY_URL;

  function handleLogin() {
    const popup = window.open(
      API_GATEWAY_URL + "/auth/oauth2/authorization/google",
      "viteUrlShortenerOauthLogin",
      "width=500,height=600"
    );

    function receiveMessage(event: MessageEvent) {
      if (event.origin !== API_GATEWAY_URL) return;

      const { token, email } = event.data;
      if (token && email) {
        dispatch(
          setCredentials({
            jwt: token,
            user: { email },
          })
        );
      }

      window.removeEventListener("message", receiveMessage);
      popup?.close();
    }

    window.addEventListener("message", receiveMessage);
  }

  return (
    <Page>
      <Title content="Login" />
      <Button label="Login with Google" onClick={handleLogin} />
    </Page>
  );
}
