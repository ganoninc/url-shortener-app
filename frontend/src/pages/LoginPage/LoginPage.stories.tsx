import type { Meta, StoryObj } from "@storybook/react-vite";
import LoginPage from "./LoginPage";
import { Provider } from "react-redux";
import { setupStore, type RootState } from "../../redux/store";

const makeStore = (preloadedState?: Partial<RootState>) =>
  setupStore(preloadedState);

const meta: Meta<typeof LoginPage> = {
  title: "Pages/LoginPage",
  component: LoginPage,
  tags: ["autodocs"],
  decorators: [
    (Story, context) => {
      const store = makeStore(context.parameters.preloadedState || {});

      return (
        <>
          <Provider store={store}>
            <Story />
          </Provider>
        </>
      );
    },
  ],
};
export default meta;

type Story = StoryObj<typeof LoginPage>;

export const Default: Story = {};
