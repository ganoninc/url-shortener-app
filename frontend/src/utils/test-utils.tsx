import { render, type RenderOptions } from "@testing-library/react";
import { setupStore, type AppStore, type RootState } from "../redux/store";
import type { PropsWithChildren, ReactNode } from "react";
import { Provider } from "react-redux";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: ReactNode,
  extendedRenderOptions: ExtendedRenderOptions = {}
) {
  const {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>{children}</Provider>
  );

  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
