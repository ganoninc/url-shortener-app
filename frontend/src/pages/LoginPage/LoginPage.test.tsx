import { screen, fireEvent } from "@testing-library/react";
import { vi } from "vitest";
import { setCredentials } from "../../redux/authSlice";
import LoginPage from "./LoginPage";
import { renderWithProviders } from "../../utils/test-utils";
import { setupStore } from "../../redux/store";

vi.stubEnv("VITE_API_GATEWAY_URL", "http://localhost");

const API_GATEWAY_URL = import.meta.env.VITE_API_GATEWAY_URL;

const jwtToken = "fake-jwt-token";
const userEmail = "user@test.com";

describe("LoginPage", () => {
  beforeAll(() => {
    console.log(window);
    Object.defineProperty(window, "open", {
      writable: true,
      value: vi.fn(),
    });

    vi.spyOn(window, "addEventListener");
    vi.spyOn(window, "removeEventListener");
  });

  afterEach(() => {
    vi.resetAllMocks();
  });

  it("matches the login page snapshot", () => {
    const { container } = renderWithProviders(<LoginPage />);
    expect(container).toMatchSnapshot();
  });

  it("opens a popup and adds message listener when login with Google", () => {
    renderWithProviders(<LoginPage />);

    fireEvent.click(screen.getByRole("button", { name: /login with google/i }));

    expect(window.open).toHaveBeenCalledTimes(1);
    expect(window.open).toHaveBeenCalledWith(
      API_GATEWAY_URL + "/auth/oauth2/authorization/google",
      "viteUrlShortenerOauthLogin",
      "width=500,height=600"
    );

    expect(window.addEventListener).toHaveBeenCalledTimes(1);
    expect(window.addEventListener).toHaveBeenCalledWith(
      "message",
      expect.any(Function)
    );
  });

  it("handles oauth callback by dispatching setCredentials", () => {
    const store = setupStore();
    vi.spyOn(store, "dispatch");
    renderWithProviders(<LoginPage />, { store });

    fireEvent.click(screen.getByRole("button", { name: /login with google/i }));
    window.dispatchEvent(
      new MessageEvent("message", {
        data: { token: jwtToken, email: userEmail },
        origin: API_GATEWAY_URL,
      })
    );

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      setCredentials({
        status: "authenticated",
        jwt: jwtToken,
        user: { email: userEmail },
      })
    );
  });

  it("closes popup and removes message listener after handling message", () => {
    const popupMock = { close: vi.fn() };
    window.open = vi.fn().mockReturnValue(popupMock);

    renderWithProviders(<LoginPage />);

    fireEvent.click(screen.getByRole("button", { name: /login with google/i }));
    const messageEvent = new MessageEvent("message", {
      data: { token: jwtToken, email: userEmail },
      origin: API_GATEWAY_URL,
    });
    window.dispatchEvent(messageEvent);

    expect(popupMock.close).toHaveBeenCalledTimes(1);
    expect(window.removeEventListener).toHaveBeenCalledTimes(1);
    expect(window.removeEventListener).toHaveBeenCalledWith(
      "message",
      expect.any(Function)
    );
  });
});
