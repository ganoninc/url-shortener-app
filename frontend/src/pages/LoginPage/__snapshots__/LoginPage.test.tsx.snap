// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > matches the login page snapshot 1`] = `
<div>
  <h1>
    Login
  </h1>
  <button>
    Login with Google
  </button>
</div>
`;
